bitcoin/src/internal_macros.rs

```diff
@@ -211,3 +211,108 @@ macro_rules! impl_asref_push_bytes {
     };
 }
 pub(crate) use impl_asref_push_bytes;
+
+/// Defines an trait `$trait_name` and implements it for `ty`, used to define extension traits.
+macro_rules! define_extension_trait {
+    // With self, no generics.
+    ($(#[$($trait_doc_attrs:tt)*])* $trait_vis:vis trait $trait_name:ident impl for $ty:ident {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
+            fn $fn:ident($slf:ident $(, $param_name:ident: $param_type:ty)*) -> $( $ret:ty )? $body:block
+        )*
+    }) => {
+        $(#[$($trait_doc_attrs)*])* $trait_vis trait $trait_name {
+            $(
+                $(#[$($fn_doc_attrs)*])*
+                fn $fn($slf $(, $param_name: $param_type)*) $( -> $ret )?;
+            )*
+        }
+
+        impl $trait_name for $ty {
+            $(
+                fn $fn($slf $(, $param_name: $param_type)*) $( -> $ret )? $body
+            )*
+        }
+    };
+    // With &self, no generics.
+    ($(#[$($trait_doc_attrs:tt)*])* $trait_vis:vis trait $trait_name:ident impl for $ty:ident {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
+            fn $fn:ident(&$slf:ident $(, $param_name:ident: $param_type:ty)*) $( -> $ret:ty )? $body:block
+        )*
+    }) => {
+        $(#[$($trait_doc_attrs)*])* $trait_vis trait $trait_name {
+            $(
+                $(#[$($fn_doc_attrs)*])*
+                fn $fn(&$slf $(, $param_name: $param_type)*) $( -> $ret )?;
+            )*
+        }
+
+        impl $trait_name for $ty {
+            $(
+                fn $fn(&$slf $(, $param_name: $param_type)*) $( -> $ret )? $body
+            )*
+        }
+    };
+    // With &self, with generics.
+    ($(#[$($trait_doc_attrs:tt)*])* $trait_vis:vis trait $trait_name:ident impl for $ty:ident {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
+            fn $fn:ident$(<$($gen:ident: $gent:ident),+>)?(&$slf:ident $(, $param_name:ident: $param_type:ty)*) $( -> $ret:ty )? $body:block
+        )*
+    }) => {
+        $(#[$($trait_doc_attrs)*])* $trait_vis trait $trait_name {
+            $(
+                $(#[$($fn_doc_attrs)*])*
+                fn $fn$(<$($gen: $gent),+>)?(&$slf $(, $param_name: $param_type)*) $( -> $ret )?;
+            )*
+        }
+
+        impl $trait_name for $ty {
+            $(
+                fn $fn$(<$($gen: $gent),+>)?(&$slf $(, $param_name: $param_type)*) $( -> $ret )? $body
```

Comment:
I understand why writing `+` is tempting but interestingly empty angle brackets are allowed by the language because it's useful for macros.

Reply:
I can change it, but just for my education, if both work what is the reason to use `*`?

Reply:
`*` allows it to be empty, which is valid Rust, `+` doesn't.

Reply:
Oh I get it, some other code might want to write `fn foo<>(self)` and `*` allows that. Cheers.

Reply:
Implemented.